nova_vm/src/ecmascript/builtins/ecmascript_function.rs

```diff
@@ -454,11 +474,76 @@ impl InternalMethods for ECMAScriptFunction {
-        _arguments_list: ArgumentsList,
-        _new_target: Function,
+        agent: &mut Agent,
+        arguments_list: ArgumentsList,
+        new_target: Function,
     ) -> JsResult<Object> {
-        todo!()
+        // 2. Let kind be F.[[ConstructorKind]].
+        let is_base = !agent[self]
+            .ecmascript_function
+            .constructor_status
+            .is_derived_class();
+        // 3. If kind is BASE, then
+        let this_argument = if is_base {
+            // a. Let thisArgument be ? OrdinaryCreateFromConstructor(newTarget, "%Object.prototype%").
+            Some(ordinary_create_from_constructor(
+                agent,
+                new_target,
+                ProtoIntrinsics::Object,
+            )?)
+        } else {
+            None
+        };
+
+        // 4. Let calleeContext be PrepareForOrdinaryCall(F, newTarget).
+        let callee_context = prepare_for_ordinary_call(agent, self, Some(new_target.into_object()));
+        // 7. Let constructorEnv be the LexicalEnvironment of calleeContext.
+        let constructor_env = callee_context
+            .ecmascript_code
+            .as_ref()
+            .unwrap()
+            .lexical_environment;
+        // 5. Assert: calleeContext is now the running execution context.
+        // assert!(std::ptr::eq(agent.running_execution_context(), callee_context));
+
+        // 6. If kind is base, then
+        if is_base {
+            // a. Perform OrdinaryCallBindThis(F, calleeContext, thisArgument).
+            ordinary_call_bind_this(
+                agent,
+                self,
+                constructor_env,
+                this_argument.unwrap().into_value(),
+            );
+            // b. Let initializeResult be Completion(InitializeInstanceElements(thisArgument, F)).
+            // c. If initializeResult is an abrupt completion, then
+            //    i. Remove calleeContext from the execution context stack and restore callerContext as the running execution context.
+            //    ii. Return ? initializeResult.
+            // TODO: Classes.
+        }
+
+        // 8. Let result be Completion(OrdinaryCallEvaluateBody(F, argumentsList)).
+        let result = ordinary_call_evaluate_body(agent, self, arguments_list);
+        // 9. Remove calleeContext from the execution context stack and restore callerContext as the running execution context.
+        agent.execution_context_stack.pop();
+        // 10. If result is a return completion, then
+        // 11. Else,
+        //   a. ReturnIfAbrupt(result).
+        let value = result?;
+        // 10. If result is a return completion, then
+        //   a. If result.[[Value]] is an Object, return result.[[Value]].
+        if value.is_object() {
+            return Ok(Object::try_from(value).unwrap());
+        }
+        //   b. If kind is base, return thisArgument.
+        if is_base {
+            return Ok(this_argument.unwrap());
+        }
+        todo!("Derived classes");
```

Comment:
**nitpick:** I guess all the steps here can already be implemented.

Reply:
Not yet. I think for derived classes, `constructor_env` would be a `FunctionEnvironment`, and `get_this_binding()` wasn't implemented yet for those. I do implement it in #197 though.